Tangy Amber Platypus

Medium

# Unrefunded excess fees occurs in `PythOracle.sol`'s `updatePrice()` function

### Summary

The` updatePrice()` function in `PythOracle.sol` accepts fees through `msg.value` but fails to refund any overpayment. As a `payable` function, excess Ether sent remains locked in the contract, which may result in losses for the caller.

### Root Cause

The `updatePrice()` function retrieves the latest price from the `PythOracle` to update the values. Being `payable`, it lets users send ETH to cover the associated update fee.

If we look at the function `updatePrice()` we can see that if the ETH sent in `msg.value` exceeds the required fee, the excess is not refunded and stays in the contract. Since there's no assurance that the next caller will be the same as the previous one, if callers accidentally send more ETH than necessary, they will lose those funds.

As this issue is related with the asset loss due to the lack of a mechanism that returns any excess Ether, I believe it's severity is Medium.

https://github.com/sherlock-audit/2024-11-oku/blob/main/oku-custom-order-types/contracts/oracle/External/PythOracle.sol#L35-L49

### Internal pre-conditions

_No response_

### External pre-conditions

User needs to pass more Ether than the update fee

### Attack Path

1. User A wants to update the price using the function `updatePrice()`
2. He puts some ETH (let's say 0.001) inside the tx
3. At the same time, the corresponding fee is 0.0002 ETH
4. The function executes correctly and results a 0.0008 ETH loss for the user

### Impact

Loss of Ether for the caller of `updatePrice()`

### PoC

_No response_

### Mitigation

```diff   
    function updatePrice(
        bytes[] calldata priceUpdate
    ) external payable override returns (uint256 updatedPrice) {
        // Submit a priceUpdate to the Pyth contract to update the on-chain price.
        // Updating the price requires paying the fee returned by getUpdateFee.
        // WARNING: These lines are required to ensure the getPriceNoOlderThan call below succeeds. If you remove them, transactions may fail with "0x19abf40e" error.
        uint fee = pythOracle.getUpdateFee(priceUpdate);
+       uint256 excess = msg.value - fee;
        pythOracle.updatePriceFeeds{value: fee}(priceUpdate);

        IPyth.Price memory price = pythOracle.getPriceNoOlderThan(
            tokenId,
            uint256(uint64(noOlderThan))
        );
        updatedPrice = uint256(uint64(price.price));
    }

+    if (excess > 0) {
+        (bool success, ) = payable(msg.sender).call{value: excess}("");
+        require(success, "Call is not successful");
    }
```